src/cargo/core/features.rs

```diff
@@ -0,0 +1,275 @@
+                $(
+                    if feature == stringify!($feature) {
+                        return Some((&mut self.$feature, stab!($stab)))
+                    }
+                )*
+                None
+            }
+        }
+    )
+}
+
+macro_rules! stab {
+    (stable) => (Status::Stable);
+    (unstable) => (Status::Unstable);
+}
+
+/// A listing of all features in Cargo
+///
+/// "look here"
+///
+/// This is the macro that lists all stable and unstable features in Cargo.
+/// You'll want to add to this macro whenever you add a feature to Cargo, also
+/// following the directions above.
+///
+/// Note that all feature names here are valid Rust identifiers, but the `_`
+/// character is translated to `-` when specified in the `cargo-features`
+/// manifest entry in `Cargo.toml`.
+features! {
+    pub struct Features {
+
+        // A dummy feature that doesn't actually gate anything, but it's used in
+        // testing to ensure that we can enable stable features.
+        [stable] test_dummy_stable: bool,
+
+        // A dummy feature that gates the usage of the `im-a-teapot` manifest
+        // entry. This is basically just intended for tests.
+        [unstable] test_dummy_unstable: bool,
+    }
+}
+
+pub struct Feature {
+    name: &'static str,
+    get: fn(&Features) -> &bool,
+}
+
+impl Features {
+    pub fn new(features: &[String],
+               warnings: &mut Vec<String>) -> CargoResult<Features> {
+        let mut ret = Features::default();
+        for feature in features {
+            ret.add(feature, warnings)?;
+            ret.activated.push(feature.to_string());
+        }
+        Ok(ret)
+    }
+
+    fn add(&mut self, feature: &str, warnings: &mut Vec<String>) -> CargoResult<()> {
+        let (slot, status) = match self.status(feature) {
+            Some(p) => p,
+            None => bail!("unknown cargo feature `{}`", feature),
+        };
+
+        if *slot {
+            bail!("the cargo feature `{}` has already bene activated", feature);
+        }
+
+        match status {
+            Status::Stable => {
+                let warning = format!("the cargo feature `{}` is now stable \
+                                       and is no longer necessary to be listed \
+                                       in the manifest", feature);
+                warnings.push(warning);
+            }
+            Status::Unstable if !nightly_features_allowed() => {
+                bail!("the cargo feature `{}` requires a nightly version of \
+                       Cargo, but this is the `{}` channel",
+                      feature,
+                      channel())
+            }
+            Status::Unstable => {}
+        }
+
+        *slot = true;
+
+        Ok(())
+    }
+
+    pub fn activated(&self) -> &[String] {
+        &self.activated
+    }
+
+    pub fn require(&self, feature: &Feature) -> CargoResult<()> {
+        if *(feature.get)(self) {
```

Comment:
`*(feature.get)(self)` looks cryptic. Why do we return `&bool`? :) 

I would try to write it like this probably:

```
if feature.is_enabled(self) {
    return Ok(())
}
// bail here
```